File: cplusplus/46Permutations.cpp
class Solution {
    vector<vector<int>> permute(vector<int>& nums, vector<bool>& selected, vector<vector<int>> &res, int pos) {
        if (pos == nums.size())
            return res;

        vector<vector<int>> output;
        for (int i = 0; i < selected.size(); ++i) {
            if (!selected[i]) {
                vector<vector<int>> list;
                selected[i] = true;
                if (pos == 0) {
                    vector<int> v;
                    v.push_back(nums[i]);
                    list.push_back(v);
                } else {
                    for (int j = 0; j < res.size(); ++j) {
                        vector<int> v = res[j];
                        v.push_back(nums[i]);
                        list.push_back(v);
                    }
                }
                
                vector<vector<int>> new_list = permute(nums, selected, list, pos + 1);
                output.insert(output.end(), new_list.begin(), new_list.end() );
                
                selected[i] = false;
            }
        }
        
        return output;
    }
    
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> v;
        if (nums.size() == 0)
            return v;
        
        vector<bool> selected(nums.size(), false);
        
        return permute(nums, selected, v, 0);
    }
};

File: java/46Permutations.java
class Solution {
    void permute(int[] nums, Set<Integer> set, List<Integer> lst, List<List<Integer>> res) {
        if (set.size() == nums.length) {
            res.add(new ArrayList<>(lst));
            return;
        }
        
        for (int i = 0; i < nums.length; ++i) {
            if (set.contains(nums[i]))
                continue;
            lst.add(nums[i]);
            set.add(nums[i]);
            permute(nums, set, lst, res);
            lst.remove(lst.size() - 1);
            set.remove(nums[i]);
        }
    } 
    
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        if (nums == null || nums.length == 0)
            return res;
        
        Set<Integer> set = new HashSet<>();
        List<Integer> lst = new ArrayList<>();
        
        permute(nums, set, lst, res);
        
        return res;
    }
}
